File: python/two-pointers/4-sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        answers = set()
        for i in range(len(nums)):
            for j in range(i + 1, len(nums)):
                left, right = j + 1, len(nums) - 1
                while left < right:
                    summation = nums[i] + nums[j] + nums[left] + nums[right]
                    if summation == target:
                        answers.add((nums[i], nums[j], nums[left], nums[right]))
                        left += 1
                        right -= 1
                    if summation > target:
                        right -= 1
                    if summation < target:
                        left += 1

        return list(answers)

File: c++/two-pointers/4-sum.cc
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        std::sort(nums.begin(), nums.end());

        std::vector<std::vector<int>> answer;

        for (int i = 0; i < nums.size(); ++i) {
            for (int j = i + 1; j < nums.size(); ++j) {
                int left = j + 1;
                int right = nums.size() - 1;

                while (left < right) {
                    long long value = 0LL + nums[i] + nums[j] + nums[left] + nums[right];

                    if (value == target) {
                        answer.push_back(std::vector<int>{nums[i], nums[j], nums[left], nums[right]});
                        left++;
                        right--;
                    } else if (value < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }

        std::sort(answer.begin(), answer.end());
        answer.erase(std::unique(answer.begin(), answer.end()), answer.end());

        return answer;
    }
};
